### PR TITLE
[imagec] introduce retry logic with backoff

### LINKED ISSUE #2015: Failed to fetch OAuth token: Get https://auth.docker.io/token?scope=repository%!A(MISSING)library%!F(MISSING)busybox%!A(MISSING)pull&service=registry.docker.io: dial tcp:
Saw this failure on a recent CI build:

[integration_logs_4080_5685722d6a53cec1a6f3a82c7d2fc5664b06a35b_1471382669.zip](https://github.com/vmware/vic/files/421318/integration_logs_4080_5685722d6a53cec1a6f3a82c7d2fc5664b06a35b_1471382669.zip)

### PR DESCRIPTION
**Pull request by caglar10ur**
Hopefully fixes #2015, #2011 and #2028

### CODE REVIEW

**Comment by hmahmood** [comment]
I am not sure this is fixing the problem. The code also looks pretty hairy, and I am nervous to just introduce it into the already unstable mess that we have.

**Comment by caglar10ur** [comment]
Not sure why you think the code is hairy, it's pretty straightforward and exactly what docker does and also I'm not sure what will be the harm. At worst case this will fail 5 times instead of giving up on 1st try.

**Comment by hmahmood** [comment]
I am a little surprised that occasional 401's from the docker registry are routine. Nonetheless, retries don't hurt.

lgtm

**Comment by caglar10ur** [comment]
I'm nearly sure that they are keeping those oauth tokens in memory (as there is no real reason to persist them) so any failure on their side will end up returning a 401 to the caller. It's just a design choice they made to make their life easier/

File: cmd/imagec/fetcher.go [hunk 1]
```diff
@@ -98,24 +101,69 @@ func NewURLFetcher(options FetcherOptions) Fetcher {
...
-	return u.fetch(ctx, url, "")
-}
-
-// FetchWithProgress fetches a web page from url and stores in a temporary file while showing a progress bar.
-func (u *URLFetcher) FetchWithProgress(url *url.URL, ID string) (string, error) {
-	defer trace.End(trace.Begin(url.String()))
+	var name string
+	var err error
+	var retries int
+	for {
```
Comment (hickeng) [inline cmd/imagec/fetcher.go]: not that I fundamentally object, but why not use the loop variables?
```
for attempts := 1; attempts < maxDownloadAttempts; attempts++ {
  ...
}
log.Debugf("Download failed: %v", err)
return "", err
```

File: cmd/imagec/fetcher.go [hunk 2]
```diff
@@ -98,24 +101,69 @@ func NewURLFetcher(options FetcherOptions) Fetcher {
...
+				}
+			case <-ctx.Done():
+				ticker.Stop()
+				return "", fmt.Errorf("download cancelled during retry delay")
+			}
+		}
+	}
 }
 
 func (u *URLFetcher) fetch(ctx context.Context, url *url.URL, ID string) (string, error) {
```
Comment (hickeng) [inline cmd/imagec/fetcher.go]: Would like a function comment explaining the behaviour with/without ID==""

**Comment by hickeng** [comment]
questions, but mostly LGTM

File: cmd/imagec/fetcher.go (hunk 1 above)
Comment (caglar10ur) [inline cmd/imagec/fetcher.go]: done

File: cmd/imagec/fetcher.go (hunk 2 above)
Comment (caglar10ur) [inline cmd/imagec/fetcher.go]: done

**Pull request merged**